### PR TITLE
🎉 Destination Redshift: File buffer increase limit

### PR DESCRIPTION
**Pull request by adam-bloom**
## What
See https://github.com/airbytehq/airbyte/pull/13975

## How
Increases the number of buffers that redshift destinations are allowed to use on disk.

## Recommended reading order
1. `x.java`
2. `y.python`

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [x] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

### CODE REVIEW

**Review by tuliren — CHANGES_REQUESTED** [review]
I was not aware of the change that introduced `maxConcurrentStreams`, which seems dangerous  (#14546). In this PR, the `maxConcurrentStreams` is set to be the same as the number of stream. This can surely lead to out-of-memory error (OOME). For example, a connection from a database can have 1000 tables / streams. One of the major reason to introduce the file-based buffer was to avoid OOME. This series of changes is kind of reverting that effort.

Maybe I missed the discussion. Will ask around.

**Comment by adam-bloom** [comment]
@tuliren how does this change the likelihood of an OOM? The overall buffer size limits remain in place - both per buffer and total across all buffers. Shouldn't those limits on their own be enough to limit memory/disk usage? Even with this change, I'm still seeing regular file buffer flushing behavior.

We've been running with this change for months since a CDC source with >10 tables is unusable otherwise - the buffering strategy will repeatedly try to open additional buffers and be forced to flush everything. Happy to share any logs from our setup that help you understand this - feel free to message me on the airbyte slack.

cc @natalyjazzviolin, as they've been helping drive this forward.

**Comment by adam-bloom** [comment]
@tuliren @natalyjazzviolin anything I can do to help move this forward? If there are reasons to not pursue this, what would be a reasonable alternative?

**Comment by marcosmarxm** [comment]
@tuliren can you check @adam-bloom comment here?

**Comment by alexnikitchuk** [comment]
@tuliren @natalyjazzviolin any chances to merge the fix? Slow replication is annoying.

**Comment by alexnikitchuk** [comment]
@tuliren looks like this gonna fix https://github.com/airbytehq/airbyte/issues/12532 as well

File: airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftStagingS3Destination.java
Comment (ChristopheDuong) [inline airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftStagingS3Destination.java:133]:
```suggestion
        CsvSerializedBuffer.createFunction(null, () -> new FileBuffer(CsvSerializedBuffer.CSV_GZ_SUFFIX)),
```
I wouldn't set it to the size of the catalog because it makes the limit useless and potentially dangerous for running out of file descriptors.

There is a more global discussion to be had, it's not just specific to Redshift: ie you could change the [constant value](https://github.com/airbytehq/airbyte/blob/b261d1fe77ab4cb4c99253cdef0e884c017507aa/airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/record_buffer/FileBuffer.java#L35) instead. 

But I agree that the limit could be set much higher than just 10, but it shouldn't be more than 10000 or something more reasonable

**Review by ChristopheDuong — CHANGES_REQUESTED** [review]
Agreed with @tuliren that the PR is not in a good state yet, but a more reasonable limit could be choosen.

(and yes, there's not a risk of OOME because of the other limits in place)

File: airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftStagingS3Destination.java
Comment (adam-bloom) [reply airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftStagingS3Destination.java:133]: I agree this is broader than just Redshift - my original PR addressed all destinations that used buffering, and then I was asked to split it up. I think we lost some context that way.

@ChristopheDuong is there a maximum number of streams supported by a CDC source? Or does bumping it up to a much larger constant just reduce the probability of encountering this issue?

Perhaps we increase the constant, but then still pass the stream count so we can log a warning about reduced performance if it exceeds? We could take the min of the value passed to the constructor or the constant.

Comment (alexnikitchuk) [reply airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftStagingS3Destination.java:133]: @adam-bloom I'd be happy if we set the constant to 100 or even make it configurable so the users can set their own value

Comment (adam-bloom) [reply airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftStagingS3Destination.java:133]: @alexnikitchuk configurable is interesting - I hadn’t considered that! I think that’s important from my comment above is helping users identify the “correct” configuration. Some sort of log that they are misconfigured may be very helpful, since currently you’d need to recognize the constant small flushes - there isn’t a clear log indicating the stream count mismatch.

Comment (alexnikitchuk) [reply airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftStagingS3Destination.java:133]: @adam-bloom yup, a warning message in the logs about performance implications would be great to have

**Comment by adam-bloom** [comment]
@alexnikitchuk @ChristopheDuong thanks for the feedback - please see the recent changes. I believe we'd still need to change this per destination that supports file buffering. I can either add BigQuery and Snowflake to this PR, or one of us can copy these changes after merging Redshift.

I just tested this out in my dev environment, and I believe the behavior is correct. Since the default is now 100, I get no log by default. As soon as I configure the value of 10, I got the following log:

```
2022-09-27 18:02:14 destination > Using destination type: COPY_S3
2022-09-27 18:02:14 destination > Potential performance issue: catalog contains 26 streams, destination configuration allocates buffers for 10 streams. If the source sends data for multiple streams in arbitrary order, this destination may be forced to flush small chunks of data repeatedly. Increase the configured File Buffer Count to improve performance
2022-09-27 18:02:14 destination > HikariPool-1 - Starting...
2022-09-27 18:02:14 destination > HikariPool-1 - Start completed.
2022-09-27 18:02:14 destination > Creating S3 client...
```

**Comment by ChristopheDuong** [comment]
@grishick do you want to take a look at this PR and discuss it with your team?

**Review by ChristopheDuong — COMMENTED** [review]
I'm deferring to the destination connector team @grishick

**Review by ChristopheDuong — APPROVED** [review]

**Comment by alexnikitchuk** [comment]
@ChristopheDuong @tuliren @marcosmarxm any updates here?

**Comment by marcosmarxm** [comment]
Hello @adam-bloom not yet, @liren is working in Redshift destination for other features and also to implement your contirbution. Soon he will return to you :)

**Comment by alexnikitchuk** [comment]
hey @marcosmarxm any ETA here?

**Comment by marcosmarxm** [comment]
Sorry the delay @adam-bloom the team is working in reviewing hacktoberfest contributions, next week we're going to return to all pending reviews!

**Comment by marcosmarxm** [comment]
Hello :wave:, first thank you for this amazing contribution. 

We really appreciate the effort you've made to improve the project.
We ask you patience for the code review. Last month our team was focused on Hacktoberfest event and that probably left some PR without the proper feedback. And this week, due to the Thanksgiving US Holiday, most our team is out of office with their families. Another important piece of information why code won't be merge this week is: as a safety measure the core team has decided to freeze merging code to main branch to keep the release stable. Next week we'll return to you with the proper code review and update the status of your contribution.

If you have any questions feel free to send me a message in Slack!
Thanks!

File: airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftStagingS3Destination.java
Comment (grishick) [inline airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftStagingS3Destination.java:138]: This line will force existing jobs to use `100` concurrent buffers instead of `10` (current behavior). To keep the changes safe for those who are already using this connector, please change the fallback value to `10`.

File: airbyte-integrations/connectors/destination-redshift/src/main/resources/spec.json
Comment (grishick) [inline airbyte-integrations/connectors/destination-redshift/src/main/resources/spec.json:215]: Current behavior is to use 10 buffers. This changes the default to 100. Let's keep it at 10 to avoid post-upgrade surprises.

File: airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftStagingS3Destination.java
Comment (grishick) [inline airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftStagingS3Destination.java:138]: nit: @adam-bloom could you please add a constant for `uploading_method` string?

**Review by grishick — CHANGES_REQUESTED** [review]
@adam-bloom thank you for the contribution. Looks good overall. I asked for two small changes in default/fallback values and a small nitpick.

File: airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftStagingS3Destination.java
Comment (adam-bloom) [reply airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftStagingS3Destination.java]: Addressed in 9d1043b

File: airbyte-integrations/connectors/destination-redshift/src/main/resources/spec.json
Comment (adam-bloom) [reply airbyte-integrations/connectors/destination-redshift/src/main/resources/spec.json]: Addressed in 9d1043b

**Review by grishick — APPROVED** [review]

File: airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftStagingS3Destination.java
Comment (grishick) [reply airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftStagingS3Destination.java]: yes. Thanks a lot!

**Comment by adam-bloom** [comment]
@grishick those failures look unrelated to this change to me (I expect those tests are also failing on `master`) - do you agree?

**Comment by marcosmarxm** [comment]
Hello :wave::skin-tone-2: and thank you for your contribution!

Airbyte has instituted a code freeze between 19 and 30 December, to make sure there are no disruptions during the holidays. 
Because of this, reviewing and merging your contribution may take longer than usual. 
We apologize for the delay, but we want everyone to have a quiet and happy holiday.

If you have any questions or need further clarification, please don't hesitate to ping via Slack.

**Review by ryankfu — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftStagingS3Destination.java
Comment (ryankfu) [reply airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftStagingS3Destination.java:144]: Following up: I've created a PR that considers your comments while also considering the concerns that Airbyte has concerning memory footprint [here](https://github.com/airbytehq/airbyte/pull/20879). Feel free to take a look, it attempts to strike the balance of giving users the ability to configure the number of file buffers while also notifying of what trade-offs are being made

Comment (adam-bloom) [reply airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftStagingS3Destination.java:144]: Hi @ryankfu - thanks for your detailed thoughts on this, and for your PR over the holiday weekend nevertheless. Your proposal makes sense to me and seems like a good compromise. Do you want to move forward with your branch/PR? Want me to pull it into this one? Either works for me!

Comment (ryankfu) [reply airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftStagingS3Destination.java:144]: Hi @adam-bloom, I can move forward with the branch I've created. I'm currently identifying the failures in the integration test which appear to be related to protocol versioning changes the platform is working on. Most of that team is out today due to holidays but am actively working on these changes on your behalf. Please bear with me until I can get the remaining integration tests resolved

I will be sure to update you once these changes have been merged and happy holidays

Comment (adam-bloom) [reply airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftStagingS3Destination.java:144]: No need to rush, and happy holidays to you as well!

Comment (ryankfu) [reply airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftStagingS3Destination.java:144]: 👋 Hi Adam, following up as the PR for configuring `destination-redshift` is in the process of publishing. Just to note, the default file buffer count is not going to be updated in favor of simply allowing a configurable value. The main concern is that while the likelihood of breaking existing users is small, it's not zero

As such, configurability is the main portion that should still unlock more for the user while we build further features that allow users to increase performance and create stability. Thanks again for your contribution and appreciate kickstarting this conversation

Comment (adam-bloom) [reply airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftStagingS3Destination.java:144]: Thanks, @ryankfu. I appreciate the thoughtfulness around how to best approach this. I'll close this PR, as I don't believe any of the changes here are needed. I will note that something similar should probably be done for bigquery and snowflake, as they also use file buffers for loading data and would theoretically experience the same performance degradation with CDC sources - I just don't use either of those destinations and haven't confirmed myself.

Comment (ryankfu) [reply airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftStagingS3Destination.java:144]: Yes, thanks for bringing up those two connectors. I will file tickets for them within our backlog to address

**Pull request closed without merging**
